hazelcast/src/main/java/com/hazelcast/map/impl/MapStoreWrapper.java

```diff
@@ -192,4 +202,29 @@ public String toString() {
         return "MapStoreWrapper{" + "mapName='" + mapName + '\''
                 + ", mapStore=" + mapStore + ", mapLoader=" + mapLoader + '}';
     }
+
+    private static class DestroyEventImpl implements TenantControl.DestroyEvent {
+        public DestroyEventImpl(String mapName) {
+            this.mapName = mapName;
+        }
+
+        @Override
+        public <TT> void destroy(TT context) {
+            HazelcastInstance instance = (HazelcastInstance)context;
+            MapService mapService = instance.getDistributedObject(MapService.SERVICE_NAME, mapName);
+            // TODO should be remove, not destroy map,
+            // but errors remain if a new class tries to load old objects,
```

Comment:
Not sure it needs to be that complicated.  I think the original intent is good enough.
If the application wants to destroy it's caches when undeployed, it can do it on it's own.

Reply:
The only way the deserialization into the new context may fail if the underlying class has changed in an incompatible way, which is not true 99% of the time.  But indeed, if it does fail, it's ok and there may or may not be need some kind of logging.
They key thing here is that the actual object references stored in Hazelcast need to be serialized and removed, during undeploy, so when re-deployed application asks for them from the cache again, they will be deserialized into the newly-loaded classes